sympy/stats/stochastic_process_types.py

```diff
@@ -883,21 +897,106 @@ def is_absorbing_chain(self):
+            ``ConditionSet`` if ``True``.
+
+        Examples
+        ========
+
+        >>> from sympy.stats import DiscreteMarkovChain
+        >>> from sympy import Matrix, S
+
+        An irreducible Markov Chain
+
+        >>> T = Matrix([[S(1)/2, S(1)/2, 0],
+        ...             [S(4)/5, S(1)/5, 0],
+        ...             [1, 0, 0]])
+        >>> X = DiscreteMarkovChain('X', trans_probs=T)
+        >>> X.stationary_distribution()
+        Matrix([[8/13, 5/13, 0]])
+
+        A reducible Markov Chain
+
+        >>> T = Matrix([[S(1)/2, S(1)/2, 0],
+        ...             [S(4)/5, S(1)/5, 0],
+        ...             [0, 0, 1]])
+        >>> X = DiscreteMarkovChain('X', trans_probs=T)
+        >>> X.stationary_distribution()
+        Matrix([[8/13 - 8*tau0/13, 5/13 - 5*tau0/13, tau0]])
+
+        >>> Y = DiscreteMarkovChain('Y')
+        >>> Y.stationary_distribution()
+        Lambda((wm, _T), Eq(wm*_T, wm))
+
+        >>> Y.stationary_distribution(condition_set=True)
+        ConditionSet(wm, Eq(wm*_T, wm))
+
+        References
+        ==========
+
+        .. [1] https://www.probabilitycourse.com/chapter11/11_2_6_stationary_and_limiting_distributions.php
+        .. [2] https://galton.uchicago.edu/~yibi/teaching/stat317/2014/Lectures/Lecture4_6up.pdf
+
+        See Also
+        ========
+
+        sympy.stats.stochastic_process_types.DiscreteMarkovChain.limiting_distribution
+        """
         trans_probs = self.transition_probabilities
-        if trans_probs == None:
-            return None
-        if isinstance(trans_probs, MatrixSymbol):
-            wm = MatrixSymbol('wm', 1, trans_probs.shape[0])
-            return Lambda((wm, trans_probs), Eq(wm*trans_probs, wm))
-        w = IndexedBase('w')
-        wi = [w[i] for i in range(trans_probs.shape[0])]
-        wm = Matrix([wi])
-        eqs = (wm*trans_probs - wm).tolist()[0]
-        eqs.append(sum(wi) - 1)
-        soln = list(linsolve(eqs, wi))[0]
+        n = self.number_of_states
+
+        if n == 0:
+            return ImmutableMatrix(Matrix([[]]))
+
+        # symbolic matrix version
+        if not self._is_numeric:
+            wm = MatrixSymbol('wm', 1, n)
+            if condition_set:
+                # fails when n is symbolic
+                # return ConditionSet(wm, And(Eq(wm*trans_probs, wm), Eq(sum(wm), 1)))
+                # fails when n is an integer
+                # return ConditionSet(wm, And(Eq(wm * trans_probs, wm), Eq(wm * OneMatrix(n, 1), Matrix([[1]]))))
+                # fails when n is symbolic
+                # return ConditionSet(wm, And(Eq(wm * trans_probs, wm), Eq(wm * ones(n, 1), Matrix([[1]]))))
+                return ConditionSet(wm, Eq(wm * trans_probs, wm))
+            else:
+                if isinstance(trans_probs, FunctionMatrix):
+                    return Lambda(wm, Eq(wm * trans_probs, wm))
+                return Lambda((wm, trans_probs), Eq(wm * trans_probs, wm))
```

Comment:
Why these two return values are different?

Reply:
The `Lambda` throws an error (I think a TypeError) when one of its arguments is a `FunctionMatrix`. This is because only symbolic-like objects can be inputs for `Lambda`. The `Lambda` was in the older version but IMHO, a `ConditionSet` is more suitable for this kind of output. This is because the definition of the fixed row vector is the vector that solves an equation.

Reply:
Maybe I should change `isinstance(trans_probs, FunctionMatrix)` to `not isinstance(trans_probs, MatrixSymbol)` because a `MatrixSymbol` is what `Lambda` is expecting.